libcnb-cargo/src/main.rs

```diff
@@ -223,3 +379,332 @@ fn calculate_dir_size(path: impl AsRef<Path>) -> io::Result<u64> {
+                    }
+                }
+            }
+
+            if !invalid_local_buildpack_uris.is_empty() {
+                let invalid_buildpack_ids: Vec<_> = invalid_local_buildpack_uris
+                    .iter()
+                    .map(|id| format!("• {id}"))
+                    .collect();
+                warn(formatdoc! { "
+                    Ignoring buildpack project from {} 
+
+                    To include this project, please fix the following URIs with invalid Buildpack Ids in the `package.toml` file:
+                    {}
+                ", &buildpack_dir.to_string_lossy(), invalid_buildpack_ids.join("\n") });
+                continue;
+            }
+        };
+
+        let buildpack_project = BuildpackProject {
+            id,
+            local_dependencies,
+            buildpack_data,
+            buildpackage_data,
+            source_dir: buildpack_dir,
+            target_dir: buildpack_target_dir,
+        };
+
+        buildpack_projects.push(buildpack_project);
+    }
+
+    buildpack_projects
+}
+
+fn get_buildpack_projects_to_compile(
+    buildpack_projects: &[BuildpackProject],
+) -> Vec<BuildpackProject> {
+    let mut buildpack_projects_seen: HashSet<&BuildpackId> = HashSet::new();
+    let mut buildpack_projects_to_compile: Vec<BuildpackProject> = vec![];
+
+    for buildpack_project in get_buildpack_projects_in_scope(buildpack_projects) {
+        for local_dependency_id in &buildpack_project.local_dependencies {
+            let local_dependency = match buildpack_projects
+                .iter()
+                .find(|buildpack_project| &buildpack_project.id == local_dependency_id)
+            {
+                Some(local_dependency) => local_dependency,
+                None => {
+                    fail_with_error(format!("Buildpack `{}` depends on local buildpack `{}` but no valid buildpack matching that id could be located in the project",
+                                    buildpack_project.id, local_dependency_id));
+                }
+            };
+
+            if !buildpack_projects_seen.contains(&local_dependency.id) {
+                buildpack_projects_seen.insert(&local_dependency.id);
+                buildpack_projects_to_compile.push(local_dependency.clone());
+            }
+        }
+        if !buildpack_projects_seen.contains(&buildpack_project.id) {
+            buildpack_projects_seen.insert(&buildpack_project.id);
+            buildpack_projects_to_compile.push(buildpack_project.clone());
+        }
+    }
+
+    buildpack_projects_to_compile
+}
+
+fn get_cargo_metadata(buildpack_dir: &Path) -> Metadata {
+    match MetadataCommand::new()
+        .manifest_path(&buildpack_dir.join("Cargo.toml"))
+        .exec()
+    {
+        Ok(cargo_metadata) => cargo_metadata,
+        Err(error) => {
+            fail_with_error(format!("Could not obtain metadata from Cargo: {error}"));
+        }
+    }
+}
+
+fn get_current_dir() -> PathBuf {
```

Comment:
again, i'm using some of these functions as required information.  i get that `current_dir` can fail but, if it does, this is an exceptional case for what we're trying to do in this process and i'd rather fail hard here because there is no alternative recovery scenario

Reply:
see latest changes